omegaconf/nodes.py

```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
         return res
 
 
+class UnionNode(ValueNode):
+    def __init__(
+        self,
+        value: Any = None,
+        key: Any = None,
+        parent: Optional[Container] = None,
+        is_optional: bool = True,
+        ref_type: Any = Union[Any],
+        element_types: List[Any] = None,  # type: ignore
+    ):
+        self.element_types = element_types
+        super().__init__(
+            parent=parent,
+            value=value,
+            metadata=Metadata(
+                key=key,
+                optional=is_optional,
+                ref_type=ref_type,
+                object_type=type(value),
+            ),
+        )
+
+    def _wrap_node(
+        self,
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[Container] = None,
+        ref_type: Any = None,
+        use_type: bool = False,
+    ) -> Optional[Node]:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        if value is None:
+            return None
+        else:
+            return _maybe_wrap(
+                ref_type=ref_type,
+                key=key,
+                value=value,
+                is_optional=is_optional,
+                parent=parent,  # type: ignore
+                use_type=use_type,
+            )
+
+    def validate_and_convert(self, value: Any) -> Optional[Any]:
+        from omegaconf._utils import is_structured_config
+        from omegaconf.basecontainer import BaseContainer
+
+        value_node = None
+        valid_value = False
+        value_type = type(value)
+        if isinstance(value, BaseContainer) and is_structured_config(
+            value._metadata.ref_type
+        ):
+            if value._metadata.ref_type in self.element_types:
+                return value
+            else:
+                self._raise_invalid_value(value)
+        if value_type in self.element_types:
```

Comment:
> (actually this would also require changing some of OmegaConf's internals, because currently typed nodes like IntegerNode, FloatNode, etc. don't support subclasses => maybe it should be part of another issue)

Yeah that's the thing with these Nodes otherwise it would be ok.

Reply:
Resolving then, @omry feel free to think about supporting subclasses more generally in typed nodes :)